momepy/preprocessing.py

```diff
@@ -772,56 +824,69 @@ def _selecting_rabs_from_poly(
-    if area_col == "area":
-        gdf.loc[:, area_col] = gdf.geometry.area
-    circom_serie = CircularCompactness(gdf, area_col).series
     # selecting roundabout polygons based on compactness
-    mask = circom_serie > circom_threshold
+    mask = gdf.reock > circom_threshold
     rab = gdf[mask]
-    # exclude those above the area threshold
+
+    # exclude those above the area threshold and larger than 2000sqmt
     area_threshold_val = gdf.area.quantile(area_threshold)
-    rab = rab[rab[area_col] < area_threshold_val]
+    if area_threshold_val > 2000:
+        rab = rab[rab[area_col] < area_threshold_val]
 
     if include_adjacent is True:
 
+        # calculating a pseudo diameter for dimeter metric later
         bounds = rab.geometry.bounds
         rab = pd.concat([rab, bounds], axis=1)
         rab["deltax"] = rab.maxx - rab.minx
         rab["deltay"] = rab.maxy - rab.miny
         rab["rab_diameter"] = rab[["deltax", "deltay"]].max(axis=1)
 
-        # selecting the adjacent areas that are of smaller than itself
+        # selecting the adjacent areas that have only three formning edges
+        gdf["savedgeom"] = gdf.geometry
         if GPD_10:
             rab_adj = gpd.sjoin(gdf, rab, predicate="intersects")
         else:
             rab_adj = gpd.sjoin(gdf, rab, op="intersects")
 
-        area_right = area_col + "_right"
-        area_left = area_col + "_left"
-        area_mask = rab_adj[area_right] >= rab_adj[area_left]
-        rab_adj = rab_adj[area_mask]
-        rab_adj.index.name = "index"
+        # remove the adjacent polygons that are selected more than once
+        rab_adj = rab_adj[~rab_adj.index.duplicated(keep=False)]
 
-        # adding a hausdorff_distance threshold
-        rab_adj["hdist"] = 0
-        # TODO: (should be a way to vectorize)
-        for i, group in rab_adj.groupby("index_right"):
-            for g in group.itertuples():
-                hdist = g.geometry.hausdorff_distance(rab.loc[i].geometry)
-                rab_adj.loc[g.Index, "hdist"] = hdist
-
-        rab_plus = rab_adj[rab_adj.hdist < (rab_adj.rab_diameter * diameter_factor)]
+        # mask based on number of forming edges (3)
+        mask_adjacents = (rab_adj["count_edges_left"] == 3) | (
+            rab_adj["reock_left"] > circom_threshold
+        )
+        rab_adj = rab_adj[mask_adjacents]
+
+        # calculating how far an adjacent polygon stretches to
+        # determine if it should still be simplified
+        max_dists = []
+        for i, row in rab_adj.iterrows():
+            xs, ys = row.geometry.exterior.coords.xy
+            poly = np.array(list(map(lambda x, y: (x, y), xs, ys)))
+
+            cx, cy = row.savedgeom.centroid.coords.xy
+            c_point = np.array(list(map(lambda x, y: (x, y), cx, cy)))
+
+            # list of distances between centroid and each vertex of polygon
+            ds = map(np.linalg.norm, c_point - poly)
+            d_max = max(ds)
+            max_dists.append(d_max)
```

Comment:
is this just a max distance between a centroid and a node? If so, you can avoid the loop and depend on shapely here for sure.

Reply:
All I could think of was this...
```python
max_dists = []
        for _, row in rab_adj.iterrows():
            dists = [
                row.savedgeom_right.centroid.distance(Point(x, y))
                for x, y in row.geometry.exterior.coords
            ]
            max_dists.append(max(dists))
        rab_adj["max_dist"] = max_dists
```
but you probably had something else in mind.